kernel/src/log_segment.rs

```diff
@@ -0,0 +1,160 @@
+//! Represents a segment of a delta log. [`LogSegment`] wraps a set of  checkpoint and commit
+//! files.
+
+use std::sync::{Arc, LazyLock};
+use url::Url;
+
+use crate::{
+    actions::{get_log_schema, Metadata, Protocol, METADATA_NAME, PROTOCOL_NAME},
+    schema::SchemaRef,
+    DeltaResult, Engine, EngineData, Error, ExpressionRef, FileMeta,
+};
+use itertools::Itertools;
+
+#[derive(Debug)]
+#[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+pub(crate) struct LogSegment {
+    pub log_root: Url,
+    /// Reverse order sorted commit files in the log segment
+    pub commit_files: Vec<FileMeta>,
+    /// checkpoint files in the log segment.
+    pub checkpoint_files: Vec<FileMeta>,
+}
+
+impl LogSegment {
+    /// Read a stream of log data from this log segment.
+    ///
+    /// The log files will be read from most recent to oldest.
+    /// The boolean flags indicates whether the data was read from
+    /// a commit file (true) or a checkpoint file (false).
+    ///
+    /// `read_schema` is the schema to read the log files with. This can be used
+    /// to project the log files to a subset of the columns.
+    ///
+    /// `meta_predicate` is an optional expression to filter the log files with. It is _NOT_ the
+    /// query's predicate, but rather a predicate for filtering log files themselves.
+    #[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+    pub(crate) fn replay(
+        &self,
+        engine: &dyn Engine,
+        commit_read_schema: SchemaRef,
+        checkpoint_read_schema: SchemaRef,
+        meta_predicate: Option<ExpressionRef>,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>> + Send> {
+        let json_client = engine.get_json_handler();
+        let commit_stream = json_client
```

Comment:
nit
```suggestion
        let commit_stream = engine.get_json_handler()
```
(again below)